### ISSUE TITLE
Speed of torch.istft

### ISSUE THREAD

**Issue by XPBooster** [pytorch/pytorch#87353]
### 🐛 Describe the bug

It is well-known that the complexity of FFT is O(NlogN) with respect to the sequence length N.
The complexity of STFT given N/T windows of length T is O(N/T * Tlog(T))~O(N).
As such, STFT should be theoretically faster than FFT given large sequence length N.

It can be backed up by the test as follow:
```
import torch
import time
device = torch.device('cuda')
length = 10000
hidden = 6
batch = 256
signal = torch.tensor([[[i for i in range(length)] for dim in range(hidden)] for b in range(batch)], device=device).float()
time_ls = []
for i in range(100):
    torch.cuda.synchronize()
    start = time.time()
    spec = torch.fft.rfft(signal)
    # torch.fft.irfft(spec)
    torch.cuda.synchronize()
    end  = time.time()
    time_ls.append(end - start)
time_ls = torch.tensor(time_ls)
time_ls = time_ls.sort()[0]
time_ls = time_ls[5:-5]
print(time_ls.mean(), time_ls.std())

time_ls = []
signal = signal.reshape(-1, length)
n_fft = 256
for i in range(100):
    torch.cuda.synchronize()
    start = time.time()
    spec = torch.stft(signal, n_fft=n_fft, hop_length=n_fft, center=False, onesided=True)
    # torch.istft(spec, n_fft=n_fft, hop_length=n_fft, center=False, onesided=True)
    torch.cuda.synchronize()
    end  = time.time()
    time_ls.append(end - start)
time_ls = torch.tensor(time_ls)
time_ls = time_ls.sort()[0]
time_ls = time_ls[5:-5]
print(time_ls.mean(), time_ls.std())
```
where the time of fft is 0.0008, and the time of stft is 0.0005.
However, if you uncomment the istft, the results are:
where the time of fft + ifft is 0.0019, and the time of stft + istft is 0.003.

Why does istft consume so much time?

**Noting that I have 

### Versions

pytorch nightly-release.
In pytorch 1.12.1 (stable version), the istft function behaves even more poorly due to #84407

**Comment by XPBooster** [comment]
cc @peterbell10

**Comment by peterbell10** [comment]
Most of the time is spent in `at::col2im`, only ~10% is the actual inverse FFT.

**Comment by XPBooster** [comment]
Yeah, I think it should be a severe bottleneck for torch.istft for large sequence.

What about the `torch.stft`? Is `at::col2im` more efficient in `torch.istft`. 
- If ture, maybe we can fix this issue and improve the efficiency.
- If not, what else hinders the speed and performance of `torch.istft`?

**Comment by XPBooster** [comment]
> Most of the time is spent in `at::col2im`, only ~10% is the actual inverse FFT.

Yeah, I think it should be a severe bottleneck for torch.istft for large sequence.

What about the `torch.stft`? Is the `at::col2im` or `::im2col` function more efficient in `torch.stft`？ 
- If ture, maybe we can fix this issue and improve the efficiency of `::col2im`.
- If not, what else hinders the speed of `torch.istft`?

**Comment by peterbell10** [comment]
`torch.stft` doesn't actually call `im2col`, it uses a stride trick to get a windowed view and then the FFT operator will use `x.contiguous()` to do the actual data movement. On the other hand `torch.istft` uses `at::col2im` which is a dedicated kernel and may not be as well optimised.

**Comment by lezcano** [comment]
Naïve question: Could we use `torch.unfold` in place of `col2im` in `istft`?

**Comment by XPBooster** [comment]
> 

Many thanks for the speedy reply. 

This problem is found in our speed comparison test and we find the strange speed of istft in large-scale benchmarks, which is inconsistent to the theoretical efficiency (faster than fft).

Is there any suggestions or possible solutions to this issue?

**Comment by XPBooster** [comment]
> `torch.stft` doesn't actually call `im2col`, it uses a stride trick to get a windowed view and then the FFT operator will use `x.contiguous()` to do the actual data movement. On the other hand `torch.istft` uses `at::col2im` which is a dedicated kernel and may not be as well optimised.

Many thanks for the speedy reply.

This problem is found in our speed comparison test and we find the strange speed of istft in large-scale benchmarks, which is inconsistent to the theoretical efficiency (faster than fft).

Is there any suggestions or possible solutions to this issue?

**Comment by XPBooster** [comment]
In addition, setting device to 'cpu', we will  find:
STFT: 0.0097s
FFT: 0.0050s

That is, running on GPUs, STFT is faster than FFT; while running on CPUs, FFT is much more faster.
It is strange since STFT has fewer flops than FFT.
Is there any comment on this issue?

**Comment by peterbell10** [comment]
stft is more or less 
```
frames = signal.unfold(1, n_fft, hop_length)
return torch.fft.rfft(frames).transpose(1, 2)
```
where `unfold` and `transpose` are views so they in theory cost nothing. However, in practice, the FFT libraries we use (cuFFT and mkl-fft) don't support arbitrary strides. So `frames` first gets copied into a new tensor that has strides the FFT libraries can work with. What you are seeing is that sometimes this copy costs more than is saved by the more efficient FFT algorithm for smaller sizes.

There is one exception, that if `frames` happens to be contiguous then there is no extra copy and stft does out-perform plain rfft. e.g. if we change `length` in your example to an even multiple of `n_fft` then I see stft as faster on CPU and CUDA:
```
In [1]: import torch
   ...: device = torch.device('cuda')
   ...: hidden = 6
   ...: batch = 256
   ...: n_fft = 256
   ...: length = 10240
   ...: signal = torch.tensor([[[i for i in range(length)] for dim in range(hidden)] for b
   ...:  in range(batch)], device=device).float().reshape(-1, length)
   ...: 
   ...: %timeit torch.fft.rfft(signal)
   ...: %timeit torch.stft(signal, n_fft=n_fft, hop_length=n_fft, center=False, return_com
   ...: plex=True, onesided=True)
   ...: s_cpu = signal.cpu()
   ...: %timeit torch.fft.rfft(s_cpu)
   ...: %timeit torch.stft(s_cpu, n_fft=n_fft, hop_length=n_fft, center=False, return_comp
   ...: lex=True, onesided=True)
1.15 ms ± 72.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
341 µs ± 939 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
8.74 ms ± 120 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
7.66 ms ± 223 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by naba89** [comment]
Hi, 

I think one of the bottlenecks for istft on GPU is the NOLA check, which causes synchronization and device-host mem copy. The istft function spends almost the entire time just that one check. On the CPU, the problem is not so severe since there is no synchornization. The problem is even worse if istft is part of a model, causes everything to synchronize and massively slows it down.

I created a naive python replica of the istft function without the NOLA check. 

Below are the profiling runs for a typical use case of batches of stereo signals of 7 seconds at 44.1 KHz (B, 2, 7*44100), without and with NOLA check.

The problem gets worse as the number of channels increases. 

B = 1 
![image](https://user-images.githubusercontent.com/12119806/218633007-d7a33a37-a514-4c08-8733-c665f37e6873.png)

B=2
![image](https://user-images.githubusercontent.com/12119806/218632553-a71d384f-0ea9-4899-8cc8-52cb8bc5dd5f.png)

B=4
![image](https://user-images.githubusercontent.com/12119806/218632604-96ba36b8-1640-4db5-959f-f28440ed3760.png)

B=8
![image](https://user-images.githubusercontent.com/12119806/218632673-c26166c4-b3b3-43b1-a3f5-34c7cf862224.png)

B=16
![image](https://user-images.githubusercontent.com/12119806/218632716-34c173ec-a383-4e93-82dc-04e36c301362.png)

I guess you get the idea, the NOLA check is hugely problematic for large batches.

I proposed to add an option to bypass the NOLA check for advanced users at  #94718
